pennylane_lightning/src/util/Kokkos_Sparse.hpp

```diff
@@ -0,0 +1,179 @@
+                                     const index_type>;
+template <class fp_precision>
+using const_graph_type =
+    typename const_crs_matrix_type<fp_precision>::staticcrsgraph_type;
+
+template <class fp_precision>
+using data_view_type =
+    typename Kokkos::View<data_type<fp_precision> *, default_layout,
+                          device_type, Kokkos::MemoryTraits<Kokkos::Unmanaged>>;
+
+template <class fp_precision>
+using const_data_view_type =
+    typename Kokkos::View<const data_type<fp_precision> *, default_layout,
+                          device_type, Kokkos::MemoryTraits<Kokkos::Unmanaged>>;
+
+/**
+ * @brief Create a Kokkos Sparse Matrix object with unmanaged views.
+ *
+ * @param row_map_ptr   Pointer to the row_map array.
+ *                      The j element encodes the number of non-zeros above row
+ * j.
+ * @param numRows       Matrix total number or rows.
+ * @param entries_ptr   Pointer to the array with the non-zero elements column
+ * indices.
+ * @param values_ptr    Pointer to the array with the non-zero elements.
+ * @param numNNZ        Number of non-zero elements.
+ * @return crs_matrix_type
+ */
+template <class fp_precision>
+const_crs_matrix_type<fp_precision> create_Kokkos_Sparse_Matrix(
+    const index_type *row_map_ptr, const index_type numRows,
+    const index_type *entries_ptr, const std::complex<fp_precision> *values_ptr,
+    const index_type numNNZ) {
+    const_index_view_type row_map(row_map_ptr, numRows + 1);
+    const_index_view_type entries(entries_ptr, numNNZ);
+    const_data_view_type<fp_precision> values(
+        reinterpret_cast<const Kokkos::complex<fp_precision> *>(values_ptr),
+        numNNZ);
+
+    const_graph_type<fp_precision> myGraph(entries, row_map);
+    const_crs_matrix_type<fp_precision> SparseMatrix("matrix", numRows, values,
+                                                     myGraph);
+    return SparseMatrix;
+}
+
+/**
+ * @brief Apply a sparse matrix to a vector.
+ *
+ * @tparam fp_precision
+ * @param vector_ptr    pointer to the vector.
+ * @param vector_size   size of the vector.
+ * @param row_map_ptr   pointer to the row_map.
+ *                      The j element encodes the number of non-zeros above row
+ * j.
+ * @param row_map_size  number of elements in the row_map.
+ * @param entries_ptr   pointer to the column indices of the non-zero elements.
+ * @param values_ptr    non-zero elements.
+ * @param numNNZ        number of non-zero elements.
+ * @return std::vector<data_type<fp_precision>>
+ */
+template <class fp_precision>
+std::vector<std::complex<fp_precision>> apply_Sparse_Matrix(
+    const std::complex<fp_precision> *vector_ptr, const index_type vector_size,
+    const index_type *row_map_ptr, const index_type row_map_size,
+    const index_type *entries_ptr, const std::complex<fp_precision> *values_ptr,
+    const index_type numNNZ) {
+
+    std::vector<std::complex<fp_precision>> result;
+    Kokkos::initialize();
```

Comment:
What kinda of setup happens in the init here? Is there much overhead? Should we move the init to the class ctor and the finalise to the destructor?

Reply:
I made some tests a while ago and it was really light.
I'm not against moving it to an upper layer, though.

Reply:
I think if there is nothing heavy in here, we can keep it in place for now